### PR TITLE
Timbelime train v2

### PR DESCRIPTION
**Pull request by timbelime**
Updated fastai v2 version of train.py

### CODE REVIEW

**Review by mnpinto — CHANGES_REQUESTED** [review]
**_This review is AI generated, it may contain errors or inaccuracies_**

# PR #14 Auto Review: `timbelime-train-v2` → `v0.7`

## Overview

This PR adds 290 lines to `banet/train.py`, porting the training pipeline from master (fastai v1) to fastai v2. The core migration is well done, but there are several bugs and behavioral changes that need attention before merging.

---

## 🔴 Bugs (Blocking)

### 1 — `satellite` parameter hardcoded as `'VIIRS375'` in `train_model()`

Lines 231–232 ignore the `satellite` function parameter and hardcode `'VIIRS375'` (which also differs from the default `'VIIRS750'`). The `target_product` parameter is similarly hardcoded.

```python
# Current (broken)
info_train = set_info_df(dsets.train.items, satellite='VIIRS375', target_product='MCD64A1C6')
info_valid = set_info_df(dsets.valid.items, satellite='VIIRS375', target_product='MCD64A1C6')

# Fix
info_train = set_info_df(dsets.train.items, satellite=satellite, target_product=target_product)
info_valid = set_info_df(dsets.valid.items, satellite=satellite, target_product=target_product)
```

### 2 — `n_episodes_train` and `n_episodes_valid` parameters ignored

The `CustomLoader` constructors hardcode `n_episodes=2000` and `n_episodes=100` instead of using the function parameters.

```python
# Current (broken)
sampler=SampleEpisode(..., n_episodes=2000, ...)   # should be n_episodes_train
sampler=SampleEpisode(..., n_episodes=100, ...)     # should be n_episodes_valid

# Fix
sampler=SampleEpisode(..., n_episodes=n_episodes_train, ...)
sampler=SampleEpisode(..., n_episodes=n_episodes_valid, ...)
```

### 3 — `pd.array()` wrapping of file list

`train_files = pd.array(sorted(...))` is non-standard and may break the `path_img/train_files` path construction. Master uses a plain list.

```python
# Fix
train_files = sorted([f.name for f in path_img.iterdir()])
```

---

## 🟡 Behavioral Differences vs. Master (Needs Confirmation)

### 4 — FRP `log1p` preprocessing removed from `open_mat()`

Master applies `np.log1p()` to the FRP channel before returning. This PR omits it. If the normalization stats (`mean[3]=0.0016`, `std[3]=0.0776`) were computed on log-transformed values, normalization will be incorrect. Was this removal intentional?

```python
# Master has this, PR does not:
data[-1, ...] = np.log1p(data[-1, ...])
data[np.isnan(data)] = 0
```

### 5 — BCE loss coefficient changed from `100*` to `50*`

The inline comment says *"use 100x for 64 days model and 50x for 32 days model"*, but the code uses `50*` while `sequence_len` defaults to 64. Either the comment or the value is wrong. Consider making it configurable:

```python
class BCE(Module):
    def __init__(self, weight=100):
        self.weight = weight
    def forward(self, x, y):
        bce = nn.BCEWithLogitsLoss()
        return self.weight * bce(x.view(x.size()[0], -1), y.view(y.size()[0], -1))
```
### 6 — `Brightness`/`Contrast` augmentation parameters differ from master

Master uses fastai v1's `get_transforms()` with `max_lighting=0.2` (the default), which applies the same lighting magnitude to both brightness and contrast. The PR splits these into `Brightness(max_lighting=0.1)` and `Contrast(max_lighting=0.2)`, so brightness is halved compared to master. Is this intentional? Note that both master and this PR apply these augmentations to all 4 channels including FRP.

---

## 🔵 Minor Issues

### 7 — Default `save_to` filename lacks `val_year`/`r_fold`

Master saves as `banet-val{val_year}-fold{r_fold}-test.pth`. The default `'banet-val-fold-v2.pth'` will overwrite on repeated runs with different folds. Suggestion:

```python
save_to=None  # in signature
...
if save_to is None:
    save_to = f'banet-val{val_year}-fold{r_fold}-v2.pth'
```

### 8 — Commented-out debug prints (lines 234–240)

Should be removed before merging.

### 9 — `# AUTOGENERATED!` header

If nbdev is still used, changes should go through the notebook. If not, update or remove the header.

---

## ✅ Positives

- fastai v2 migration is well-structured (`Callback`/`before_batch`, `RandTransform`, `DataLoaders`)
- `SampleEpisode.random_sample()` is simpler and more correct than master's two-step look-ahead
- Pretrained weights loading is more robust (handles `weights_only=False` and both dict/raw formats)
- `mae` metric correctly drops deprecated `.byte()` indexing
- `CutoutCombined` cleanly consolidates two separate cutout transforms

---

**Recommendation:** Request changes. Bugs #1–#3 should block merging. #4–#6 need explicit confirmation of intent.

File: banet/train.py [hunk 1]
```diff
@@ -1,3 +1,284 @@
...
+        loss_func=BCE(), wd=1e-2, metrics=[accuracy, dice2d, mae])
+
+    if get_learner: return learn
+
+    print('Starting traning loop\n')
+    learn.fit_one_cycle(n_epochs, lr*64/sequence_len)
+
+    model_path.mkdir(exist_ok=True)
+    if save_to is None:
+        save_to='banet-val{val_year}-fold{r_fold}-v2.pth'
```
Comment (mnpinto) [inline banet/train.py]: Missing f prefix

```python
save_to=f'banet-val{val_year}-fold{r_fold}-v2.pth'
```

File: banet/train.py [hunk 2]
```diff
@@ -1,3 +1,284 @@
...
+    learn = Learner(dls, model, cbs=[
+        ImageSequence(sequence_len=sequence_len, n_sequences=n_sequences),
+        GradientClip(1.0)
+        ],
+        loss_func=BCE(), wd=1e-2, metrics=[accuracy, dice2d, mae])
+
+    if get_learner: return learn
+
+    print('Starting traning loop\n')
+    learn.fit_one_cycle(n_epochs, lr*64/sequence_len)
```
Comment (mnpinto) [inline banet/train.py]: For the learning rate I think should be `lr * sequence_len / 64` so that smaller sequences use a smaller learning rate

File: banet/train.py [hunk 3]
```diff
@@ -1,3 +1,284 @@
...
+    valid_idx = train_df.loc[train_df.times.dt.year == val_year].index.values
+
+    dblock = DataBlock(
+        blocks=(TransformBlock, TransformBlock),
+        get_x= lambda x: TensorImage(open_mat(x)),
+        get_y= lambda x: TensorBase(open_mask(get_y_fn(str(x), satellite, target_product))),
+        splitter=IndexSplitter(valid_idx)
+    )
+
+    dsets = dblock.datasets(path_img/train_files)
```
Comment (mnpinto) [inline banet/train.py]: Here `train_files` is a list, I think `path_img/train_files` will fail, you can do
```python
dsets = dblock.datasets([path_img/f for f in train_files])
```

File: banet/train.py [hunk 4]
```diff
@@ -1,3 +1,284 @@
...
+        return list(self.custom_sampler)
+
+class ImageSequence(Callback):
+    def __init__(self, sequence_len=64, n_sequences=1):
+        self.sequence_len = sequence_len
+        self.n_sequences = n_sequences
+
+    def before_batch(self):
+        x, y = self.learn.xb[0], self.learn.yb[0]
+#        print(f"Before reshape: x={x.shape}, y={y.shape}, dtype={x.dtype}, mu_dtype={self.mean.dtype}")  # Debug
```
Comment (mnpinto) [inline banet/train.py]: Can remove debug comment

File: banet/train.py [hunk 5]
```diff
@@ -1,3 +1,284 @@
...
+        bs, ch, sz1, sz2 = x.shape
+        x = x.view(self.sequence_len, self.n_sequences, ch, sz1, sz2).permute(1, 2, 0, 3, 4)
+        y = y.view(self.sequence_len, self.n_sequences, 1, sz1, sz2).permute(1, 2, 0, 3, 4)
+#        print(f"After reshape: x={x.shape}, y={y.shape}")  # Debug
```
Comment (mnpinto) [inline banet/train.py]: Can remove debug comment

**Review by mnpinto — CHANGES_REQUESTED** [review]

**Review by mnpinto — APPROVED** [review]
